Failures mirroring files in dev catalog `dcp3`

Two files failed to mirror from the `dev` catalog `dcp3`:
- File 089b5277 failed due to an unexpected size
- File eb01f915 failed due to "File object is already present"

Failed messages were fed back to the mirror queue, however again failed with the same errors.

## Failed messages
```
{
    "queue": "https://sqs.us-east-1.amazonaws.com/122796619775/azul-mirror_fail-dev.fifo",
    "messages": [
        {
            "MessageId": "46eadd27-4aa9-4aff-94ee-a445aa2f2f2f",
            "ReceiptHandle": "AQEBZAnoZf1IQ6b/v4S/CqGpmtaZu8uoG42ruphs0KVq9TfpXw1DSRfXxv/IHYKSaDEJXYyEsiDIjL7u2f/DYIkdBYhWnh5aMzjrS7/nCDb5qgMmNfhbUpIZaVg6reprt+Bm9CiD7h7Sk+PHbwlq3xhkzIFspolWVEgBCQ2g+Yh7xhtAczpt5mcJjFLOWWKx+vkGZToWrn4hAJtO61j6Qfzx+rcMfni2yb02sJkHca1Y8lQgRJbMk/GsUABMIee1p6pq5ndNMi/QRFYO70rUso+Q8vfpViZmUl3Ovk2OWI4ANMs=",
            "MD5OfBody": "87ae1989001e27a2e6fbc6506c39119a",
            "Body": {
                "catalog": "dcp3",
                "source": {
                    "id": "251c52e6-3bda-48df-a4f2-50aa759d7d80",
                    "spec": "tdr:bigquery:gcp:datarepo-dev-e8e0a59a:hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903",
                    "prefix": "/0",
                    "type": "azul.terra.TDRSourceRef"
                },
                "prefix": "",
                "file": {
                    "uuid": "089b5277-6525-4c3f-8337-0d52352b7f51",
                    "name": "SRR5295225_1.fastq.gz",
                    "version": "2019-07-26T18:23:16.057882Z",
                    "drs_uri": "drs://jade.datarepo-dev.broadinstitute.org/v1_251c52e6-3bda-48df-a4f2-50aa759d7d80_8f559d9f-155f-4788-9354-094921708caa",
                    "size": 73473013,
                    "content_type": "application/gzip",
                    "sha256": "529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80",
                    "crc32c": "e5b3c2b1",
                    "sha1": "b2bfd1cfd16e7ea6ae6cdf8e00d89bcce5fcfb0e",
                    "s3_etag": "84044d772cdd35c60a290f0a690ca2ee-2",
                    "type": "azul.plugins.metadata.hca.HCAFile"
                },
                "action": "MirrorFileAction"
            },
            "Attributes": {
                "SenderId": "AROARZFZ7W77VNWF46R45:azul-indexer-dev-mirror",
                "ApproximateFirstReceiveTimestamp": "1765593561945",
                "ApproximateReceiveCount": "11",
                "SentTimestamp": "1765579139474",
                "SequenceNumber": "74238980951278526720",
                "MessageDeduplicationId": "46eadd27-4aa9-4aff-94ee-a445aa2f2f2f",
                "MessageGroupId": "529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80",
                "AWSTraceHeader": "Root=1-693c997e-e936866d8ba5dfc57114405b;Parent=56d52c8a80ea4f99;Sampled=0;Lineage=2:60ede013:1",
                "DeadLetterQueueSourceArn": "arn:aws:sqs:us-east-1:122796619775:azul-mirror-dev.fifo"
            },
            "_Attributes": {
                "SentTimestamp": "2025-12-12T14:38:59.474000-08:00",
                "ApproximateFirstReceiveTimestamp": "2025-12-12T18:39:21.945000-08:00"
            }
        },
        {
            "MessageId": "43044f32-3407-482b-af70-4b46e574be5c",
            "ReceiptHandle": "AQEBnSVSCP8slM/T1xrXib9Fxx8pdZb9hgrV5aoGf+WtzLUzOtAH2Wo1nB/kqC/+23trBcq8kNe70xAEXKHNJtDcEj0pU2Y3MlnIQXdOx4+G/95Tdosb4h47NlAS0CtQ9J3Nsj1z5/fkF4drKilUQ8oH6TvJFy5jG1Zj+HBRtgiTZc2l3Pao70rlP6jx1/7J5yZY3Zm9o7PhunqSTda5A9ZeP/ozrljcqqgU3Qggw4KCJEXXYYp858zVwKhqIIJ/XpFzFOirkmrcCgn+G0IMIL7WmPM0N+PBiqx+3Ve/oLgd4C8=",
            "MD5OfBody": "7e5dbcb985b16bfe2a23aa0db4335571",
            "Body": {
                "catalog": "dcp3",
                "source": {
                    "id": "2d6116af-ceb1-4324-ba47-65fbc4202028",
                    "spec": "tdr:bigquery:gcp:datarepo-dev-0c5c20b5:hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907",
                    "prefix": "/1",
                    "type": "azul.terra.TDRSourceRef"
                },
                "prefix": "2",
                "file": {
                    "uuid": "eb01f915-8f3d-45ab-833c-655020001d37",
                    "name": "22028_5#9_2.fastq.gz",
                    "version": "2019-05-16T05:31:49.670271Z",
                    "drs_uri": "drs://jade.datarepo-dev.broadinstitute.org/v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104",
                    "size": 32129,
                    "content_type": "application/gzip",
                    "sha256": "2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c",
                    "crc32c": "0f9eee22",
                    "sha1": "b33d207dfe232ca736ecfbcf153140424fe9926a",
                    "s3_etag": "5eac6ec6ae59b910bdcc931a6fb4f208",
                    "type": "azul.plugins.metadata.hca.HCAFile"
                },
                "action": "MirrorFileAction"
            },
            "Attributes": {
                "SenderId": "AROARZFZ7W77VNWF46R45:azul-indexer-dev-mirror",
                "ApproximateFirstReceiveTimestamp": "1765602812472",
                "ApproximateReceiveCount": "11",
                "SentTimestamp": "1765579152350",
                "SequenceNumber": "74238980952946622464",
                "MessageDeduplicationId": "43044f32-3407-482b-af70-4b46e574be5c",
                "MessageGroupId": "2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c",
                "AWSTraceHeader": "Root=1-693c998b-476779430d48e19fb16e430a;Parent=2cfcfc34e8e75d9c;Sampled=0;Lineage=2:60ede013:1",
                "DeadLetterQueueSourceArn": "arn:aws:sqs:us-east-1:122796619775:azul-mirror-dev.fifo"
            },
            "_Attributes": {
                "SentTimestamp": "2025-12-12T14:39:12.350000-08:00",
                "ApproximateFirstReceiveTimestamp": "2025-12-12T21:13:32.472000-08:00"
            }
        }
    ]
}
```

## File 089b5277-6525-4c3f-8337-0d52352b7f51


```
# /aws/lambda/azul-indexer-dev-mirror

[
    {
        "@timestamp": "2025-12-13 16:23:19.117",
        "@message": "START RequestId: 0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752 Version: 30\n"
    },
    {
        "@timestamp": "2025-12-13 16:23:19.118",
        "@message": "[INFO]\t2025-12-13T16:23:19.118Z\t0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752\tazul.indexer.action_controller\tWorker handling message {'catalog': 'dcp3', 'source': {'id': '251c52e6-3bda-48df-a4f2-50aa759d7d80', 'spec': 'tdr:bigquery:gcp:datarepo-dev-e8e0a59a:hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903', 'prefix': '/0', 'type': 'azul.terra.TDRSourceRef'}, 'prefix': '', 'file': {'uuid': '089b5277-6525-4c3f-8337-0d52352b7f51', 'name': 'SRR5295225_1.fastq.gz', 'version': '2019-07-26T18:23:16.057882Z', 'drs_uri': 'drs://jade.datarepo-dev.broadinstitute.org/v1_251c52e6-3bda-48df-a4f2-50aa759d7d80_8f559d9f-155f-4788-9354-094921708caa', 'size': 73473013, 'content_type': 'application/gzip', 'sha256': '529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80', 'crc32c': 'e5b3c2b1', 'sha1': 'b2bfd1cfd16e7ea6ae6cdf8e00d89bcce5fcfb0e', 's3_etag': '84044d772cdd35c60a290f0a690ca2ee-2', 'type': 'azul.plugins.metadata.hca.HCAFile'}, 'action': 'MirrorFileAction'}, attempt #10 (approx), message ID 46eadd27-4aa9-4aff-94ee-a445aa2f2f2f, group ID 529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80"
    },
...
    {
        "@timestamp": "2025-12-13 16:23:21.929",
        "@message": "[INFO]\t2025-12-13T16:23:21.929Z\t0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752\tazul.indexer.mirror_file_service\tDownloaded 79546647 bytes in 0.833s from file HCAFile(uuid='089b5277-6525-4c3f-8337-0d52352b7f51', name='SRR5295225_1.fastq.gz', version='2019-07-26T18:23:16.057882Z', drs_uri='drs://jade.datarepo-dev.broadinstitute.org/v1_251c52e6-3bda-48df-a4f2-50aa759d7d80_8f559d9f-155f-4788-9354-094921708caa', size=73473013, content_type='application/gzip', sha256='529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80', crc32c='e5b3c2b1', sha1='b2bfd1cfd16e7ea6ae6cdf8e00d89bcce5fcfb0e', s3_etag='84044d772cdd35c60a290f0a690ca2ee-2')"
    },
    {
        "@timestamp": "2025-12-13 16:23:21.931",
        "@message": "[WARNING]  2025-12-13T16:23:21.929Z    0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752    azul.indexer.action_controller  Worker failed to handle message SQSFifoMessage(body={'catalog': 'dcp3', 'source': {'id': '251c52e6-3bda-48df-a4f2-50aa759d7d80', 'spec': 'tdr:bigquery:gcp:datarepo-dev-e8e0a59a:hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903', 'prefix': '/0', 'type': 'azul.terra.TDRSourceRef'}, 'prefix': '', 'file': {'uuid': '089b5277-6525-4c3f-8337-0d52352b7f51', 'name': 'SRR5295225_1.fastq.gz', 'version': '2019-07-26T18:23:16.057882Z', 'drs_uri': 'drs://jade.datarepo-dev.broadinstitute.org/v1_251c52e6-3bda-48df-a4f2-50aa759d7d80_8f559d9f-155f-4788-9354-094921708caa', 'size': 73473013, 'content_type': 'application/gzip', 'sha256': '529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80', 'crc32c': 'e5b3c2b1', 'sha1': 'b2bfd1cfd16e7ea6ae6cdf8e00d89bcce5fcfb0e', 's3_etag': '84044d772cdd35c60a290f0a690ca2ee-2', 'type': 'azul.plugins.metadata.hca.HCAFile'}, 'action': 'MirrorFileAction'}, attempts=10, id='46eadd27-4aa9-4aff-94ee-a445aa2f2f2f', group_id='529c56a6fb1c7e62f70e4c18be392784bd49b3b99c71a9ca7794c3a134243b80', dedup_id='1a6735bd-2c11-4830-a8ef-0a7139277ebf')
Traceback (most recent call last):
  File "/var/task/azul/indexer/action_controller.py", line 57, in _handle_events
    message_handler(action)
    ~~~~~~~~~~~~~~~^^^^^^^^
  File "/var/task/azul/indexer/mirror_service.py", line 211, in mirror
    self._queue_messages(self._mirror(action))
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/mirror_service.py", line 184, in _queue_messages
    return self._queues.send_messages(self._mirror_queue(),
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
                                      map(MirrorAction.to_sqs, messages),
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                      rate_limit=rate_limit)
                                      ^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/queues.py", line 201, in send_messages
    for batch in chunked(messages, self.batch_size):
                 ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/more_itertools/recipes.py", line 133, in take
    return list(islice(iterable, n))
  File "/var/task/azul/indexer/mirror_service.py", line 279, in _
    service.mirror_file(a.file)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^
  File "/var/task/azul/indexer/mirror_file_service.py", line 292, in mirror_file
    file_content = self._download(file)
  File "/var/task/azul/indexer/mirror_file_service.py", line 392, in _download
    assert actual_size == size, R(f'Expected {size} bytes, got {actual_size}')
           ^^^^^^^^^^^^^^^^^^^
AssertionError: R('Expected 73473013 bytes, got 79546647')"
    },
    {
        "@timestamp": "2025-12-13 16:23:21.934",
        "@message": "END RequestId: 0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752\n"
    },
    {
        "@timestamp": "2025-12-13 16:23:21.934",
        "@message": "REPORT RequestId: 0a93cccd-47ee-5fdb-adc4-ad7ad2cd6752\tDuration: 2816.19 ms\tBilled Duration: 2817 ms\tMemory Size: 512 MB\tMax Memory Used: 469 MB\t\n"
    }
]
```

## File eb01f915-8f3d-45ab-833c-655020001d37

```
# /aws/lambda/azul-indexer-dev-mirror

[
    {
        "@timestamp": "2025-12-13 16:23:25.630",
        "@message": "START RequestId: 3a91fdfc-cfa0-50a2-9fb9-9c4b5a895f2b Version: 30\n"
    },
    {
        "@timestamp": "2025-12-13 16:23:25.631",
        "@message": "[INFO]\t2025-12-13T16:23:25.631Z\t3a91fdfc-cfa0-50a2-9fb9-9c4b5a895f2b\tazul.indexer.action_controller\tWorker handling message {'catalog': 'dcp3', 'source': {'id': '2d6116af-ceb1-4324-ba47-65fbc4202028', 'spec': 'tdr:bigquery:gcp:datarepo-dev-0c5c20b5:hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907', 'prefix': '/1', 'type': 'azul.terra.TDRSourceRef'}, 'prefix': '2', 'file': {'uuid': 'eb01f915-8f3d-45ab-833c-655020001d37', 'name': '22028_5#9_2.fastq.gz', 'version': '2019-05-16T05:31:49.670271Z', 'drs_uri': 'drs://jade.datarepo-dev.broadinstitute.org/v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104', 'size': 32129, 'content_type': 'application/gzip', 'sha256': '2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c', 'crc32c': '0f9eee22', 'sha1': 'b33d207dfe232ca736ecfbcf153140424fe9926a', 's3_etag': '5eac6ec6ae59b910bdcc931a6fb4f208', 'type': 'azul.plugins.metadata.hca.HCAFile'}, 'action': 'MirrorFileAction'}, attempt #10 (approx), message ID 43044f32-3407-482b-af70-4b46e574be5c, group ID 2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c"
    },
...
    {
        "@timestamp": "2025-12-13 16:23:25.720",
        "@message": "[WARNING]  2025-12-13T16:23:25.719Z    3a91fdfc-cfa0-50a2-9fb9-9c4b5a895f2b    azul.indexer.action_controller  Worker failed to handle message SQSFifoMessage(body={'catalog': 'dcp3', 'source': {'id': '2d6116af-ceb1-4324-ba47-65fbc4202028', 'spec': 'tdr:bigquery:gcp:datarepo-dev-0c5c20b5:hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907', 'prefix': '/1', 'type': 'azul.terra.TDRSourceRef'}, 'prefix': '2', 'file': {'uuid': 'eb01f915-8f3d-45ab-833c-655020001d37', 'name': '22028_5#9_2.fastq.gz', 'version': '2019-05-16T05:31:49.670271Z', 'drs_uri': 'drs://jade.datarepo-dev.broadinstitute.org/v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104', 'size': 32129, 'content_type': 'application/gzip', 'sha256': '2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c', 'crc32c': '0f9eee22', 'sha1': 'b33d207dfe232ca736ecfbcf153140424fe9926a', 's3_etag': '5eac6ec6ae59b910bdcc931a6fb4f208', 'type': 'azul.plugins.metadata.hca.HCAFile'}, 'action': 'MirrorFileAction'}, attempts=10, id='43044f32-3407-482b-af70-4b46e574be5c', group_id='2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c', dedup_id='a649f1c8-bf53-452b-93a4-cb77cef52820')
Traceback (most recent call last):
  File "/var/task/azul/indexer/action_controller.py", line 57, in _handle_events
    message_handler(action)
    ~~~~~~~~~~~~~~~^^^^^^^^
  File "/var/task/azul/indexer/mirror_service.py", line 211, in mirror
    self._queue_messages(self._mirror(action))
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/mirror_service.py", line 184, in _queue_messages
    return self._queues.send_messages(self._mirror_queue(),
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
                                      map(MirrorAction.to_sqs, messages),
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                      rate_limit=rate_limit)
                                      ^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/queues.py", line 201, in send_messages
    for batch in chunked(messages, self.batch_size):
                 ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/more_itertools/recipes.py", line 133, in take
    return list(islice(iterable, n))
  File "/var/task/azul/indexer/mirror_service.py", line 274, in _
    assert False, R('File object is already present', a.file)
           ^^^^^
AssertionError: R('File object is already present', HCAFile(uuid='eb01f915-8f3d-45ab-833c-655020001d37', name='22028_5#9_2.fastq.gz', version='2019-05-16T05:31:49.670271Z', drs_uri='drs://jade.datarepo-dev.broadinstitute.org/v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104', size=32129, content_type='application/gzip', sha256='2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c', crc32c='0f9eee22', sha1='b33d207dfe232ca736ecfbcf153140424fe9926a', s3_etag='5eac6ec6ae59b910bdcc931a6fb4f208'))"
    },
    {
        "@timestamp": "2025-12-13 16:23:25.723",
        "@message": "END RequestId: 3a91fdfc-cfa0-50a2-9fb9-9c4b5a895f2b\n"
    },
    {
        "@timestamp": "2025-12-13 16:23:25.723",
        "@message": "REPORT RequestId: 3a91fdfc-cfa0-50a2-9fb9-9c4b5a895f2b\tDuration: 92.47 ms\tBilled Duration: 93 ms\tMemory Size: 512 MB\tMax Memory Used: 469 MB\t\n"
    }
]



[
    {
        "@timestamp": "2025-12-13 16:23:25.721",
        "@message": "[ERROR] AssertionError: R('File object is already present', HCAFile(uuid='eb01f915-8f3d-45ab-833c-655020001d37', name='22028_5#9_2.fastq.gz', version='2019-05-16T05:31:49.670271Z', drs_uri='drs://jade.datarepo-dev.broadinstitute.org/v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104', size=32129, content_type='application/gzip', sha256='2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c', crc32c='0f9eee22', sha1='b33d207dfe232ca736ecfbcf153140424fe9926a', s3_etag='5eac6ec6ae59b910bdcc931a6fb4f208'))
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 194, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1761, in __call__
    return self.handler(event_obj)
  File "/var/task/azul/indexer/mirror_controller.py", line 63, in mirror
    self.mirror(event)
  File "/var/task/azul/indexer/mirror_controller.py", line 69, in mirror
    self._handle_events(event, self.service.mirror)
  File "/var/task/azul/indexer/action_controller.py", line 57, in _handle_events
    message_handler(action)
  File "/var/task/azul/indexer/mirror_service.py", line 211, in mirror
    self._queue_messages(self._mirror(action))
  File "/var/task/azul/indexer/mirror_service.py", line 184, in _queue_messages
    return self._queues.send_messages(self._mirror_queue(),
  File "/var/task/azul/queues.py", line 201, in send_messages
    for batch in chunked(messages, self.batch_size):
  File "/opt/python/more_itertools/recipes.py", line 133, in take
    return list(islice(iterable, n))
  File "/var/task/azul/indexer/mirror_service.py", line 274, in _
    assert False, R('File object is already present', a.file)"
    }
]
```

# File eb01f915-8f3d-45ab-833c-655020001d37

Again, the metadata in the snapshot does not match the downloaded file object. (Side note: I suspect there was a mix-up between the compressed and uncompressed versions of the file, since the format is declared as "fastq.gz" but the downloaded file is not a zip archive but a plain text file, and 256x times larger than declared.).

```
select 
  json_extract_scalar(descriptor, "$.size") as file_size,  
  json_extract_scalar(descriptor, '$.sha256') as sha256
from `datarepo-dev-0c5c20b5.hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907.sequence_file`
where json_extract_scalar(descriptor, '$.file_id') = 'eb01f915-8f3d-45ab-833c-655020001d37'
```
```
[{
  "file_size": "32129",
  "sha256": "2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c"
}]
```
```
$ curl -L 'https://service.dev.singlecell.gi.ucsc.edu/repository/files/eb01f915-8f3d-45ab-833c-655020001d37?catalog=dcp3&version=2019-05-16T05%3A31%3A49.670271Z' -o download.fastq.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
100 8048k  100 8048k    0     0  2438k      0  0:00:03  0:00:03 --:--:-- 8601k
$ du -b download.fastq.gz 
8241718	download.fastq.gz
$ sha256sum download.fastq.gz 
6c5a2b6e84e906feb6530ff9835a6ef6aa7bbe035d6f092c7d65db8a697efcee  download.fastq.gz
```

The error message looks different because we already tried and failed to mirror this file earlier this year and the mirror bucket was left in an inconsistent state. [A file object](https://us-east-1.console.aws.amazon.com/s3/object/edu-ucsc-gi-platform-hca-dev-mirror-dev.us-east-1?region=us-east-1&prefix=file/2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c.sha256) was created for this file on (May 27th), however no corresponding info object is present. Presumably, none was ever created, because CloudWatch shows that mirroring the file originally failed because its digest did not match the expected value. At this time, we did not separately verify the file size.

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-dev-mirror    
data-sources:     
facets:     
start-time: 2025-05-26T07:00:00.000Z    
end-time: 2025-05-29T06:59:59.000Z    
query-string:
  ```
  fields @message
| filter @message like 'eb01f915-8f3d-45ab-833c-655020001d37' and @message like 'AssertionError'
  ```
---
| @message |
| --- |
| [ERROR] AssertionError: R(&#x27;File digest value does not match its contents&#x27;, Digest(type=&#x27;sha256&#x27;, value=&#x27;2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c&#x27;), HCAFile(uuid=&#x27;eb01f915-8f3d-45ab-833c-655020001d37&#x27;, name=&#x27;22028_5#9_2.fastq.gz&#x27;, version=&#x27;2019-05-16T05:31:49.670271Z&#x27;, drs_uri=&#x27;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v1_2d6116af-ceb1-4324-ba47-65fbc4202028_ea9fce94-c8ef-48fc-9648-652a2f92b104&#x27;, size=32129, content_type=&#x27;application&#x2f;gzip&#x27;, sha256=&#x27;2bc19029cf3d4ccc45c9834851ce694a1d83c92d1d8ba173533574762239671c&#x27;, crc32c=&#x27;0f9eee22&#x27;, sha1=&#x27;b33d207dfe232ca736ecfbcf153140424fe9926a&#x27;, s3_etag=&#x27;5eac6ec6ae59b910bdcc931a6fb4f208&#x27;)) Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 187, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 289, in mirror     app.mirror_controller.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 76, in mirror     self._handle_events(event, self._mirror)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 63, in _handle_events     message_handler(message)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 88, in _mirror     self.mirror_file(json_str(message[&#x27;catalog&#x27;]),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 162, in mirror_file     self.service.mirror_file(catalog, file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 226, in mirror_file     self._verify_digest(file, hasher)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 344, in _verify_digest     assert expected_digest.value == actual_digest_value, R( |

--- a/UPGRADING.rst
+++ b/UPGRADING.rst
@@ -20,6 +20,15 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#7646 Failures mirroring files in dev catalog dcp3
+==================================================
+
+Add the ``no_mirror`` flag to the ``f8aa201c`` and ``8c3c290d`` sources in the
+``dcp3`` catalog in the ``environment.py`` files for each of your personal
+deployments. As always, use the sandbox deployment's ``environment.py`` as a
+model when upgrading personal deployments.
+
+
 #8075 Remove snapshots from dev to reduce storage to ≤ 5 TiB
 ============================================================
 

--- a/deployments/dev/environment.py
+++ b/deployments/dev/environment.py
@@ -122,7 +122,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-bd995e95', 'hca_dev_842605c7375a47c59e2ca71c2c00fcad__20210830_20210903'),
     source('bigquery', 'datarepo-dev-d0772077', 'hca_dev_8787c23889ef4636a57d3167e8b54a80__20210827_20210903'),
     source('bigquery', 'datarepo-dev-8eb2ffd1', 'hca_dev_87d52a86bdc7440cb84d170f7dc346d9__20210830_20210903'),
-    source('bigquery', 'datarepo-dev-0c5c20b5', 'hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907'),
+    source('bigquery', 'datarepo-dev-0c5c20b5', 'hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907', no_mirror),  # noqa: E501, contains file with bad size and sha256
     source('bigquery', 'datarepo-dev-29509483', 'hca_dev_90bd693340c048d48d76778c103bf545__20210827_20211110'),
     source('bigquery', 'datarepo-dev-59d37b9a', 'hca_dev_946c5add47d1402a97bba5af97e8bce7__20210831_20210903'),
     source('bigquery', 'datarepo-dev-788c3b52', 'hca_dev_955dfc2ca8c64d04aa4d907610545d11__20210831_20210903', no_mirror),  # noqa: E501, all files above mirror size limit
@@ -166,7 +166,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-10f0610a', 'hca_dev_f81efc039f564354aabb6ce819c3d414__20210827_20210903'),
     source('bigquery', 'datarepo-dev-24e9529e', 'hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908'),
     source('bigquery', 'datarepo-dev-67240cf2', 'hca_dev_f86f1ab41fbb4510ae353ffd752d4dfc__20210901_20210903'),
-    source('bigquery', 'datarepo-dev-e8e0a59a', 'hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903'),
+    source('bigquery', 'datarepo-dev-e8e0a59a', 'hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903', no_mirror),  # noqa: E501, contains file with bad size and sha256
     source('bigquery', 'datarepo-dev-96d8e08c', 'hca_dev_faeedcb0e0464be7b1ad80a3eeabb066__20210831_20210903'),
     # @formatter:on
 ]))

--- a/deployments/sandbox/environment.py
+++ b/deployments/sandbox/environment.py
@@ -124,7 +124,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-bd995e95', 'hca_dev_842605c7375a47c59e2ca71c2c00fcad__20210830_20210903'),
     source('bigquery', 'datarepo-dev-d0772077', 'hca_dev_8787c23889ef4636a57d3167e8b54a80__20210827_20210903'),
     source('bigquery', 'datarepo-dev-8eb2ffd1', 'hca_dev_87d52a86bdc7440cb84d170f7dc346d9__20210830_20210903'),
-    source('bigquery', 'datarepo-dev-0c5c20b5', 'hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907'),
+    source('bigquery', 'datarepo-dev-0c5c20b5', 'hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907', no_mirror),  # noqa: E501, contains file with bad size and sha256
     source('bigquery', 'datarepo-dev-29509483', 'hca_dev_90bd693340c048d48d76778c103bf545__20210827_20211110'),
     source('bigquery', 'datarepo-dev-59d37b9a', 'hca_dev_946c5add47d1402a97bba5af97e8bce7__20210831_20210903'),
     source('bigquery', 'datarepo-dev-788c3b52', 'hca_dev_955dfc2ca8c64d04aa4d907610545d11__20210831_20210903', no_mirror),  # noqa: E501, all files above mirror size limit
@@ -168,7 +168,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-10f0610a', 'hca_dev_f81efc039f564354aabb6ce819c3d414__20210827_20210903'),
     source('bigquery', 'datarepo-dev-24e9529e', 'hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908'),
     source('bigquery', 'datarepo-dev-67240cf2', 'hca_dev_f86f1ab41fbb4510ae353ffd752d4dfc__20210901_20210903'),
-    source('bigquery', 'datarepo-dev-e8e0a59a', 'hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903'),
+    source('bigquery', 'datarepo-dev-e8e0a59a', 'hca_dev_f8aa201c4ff145a4890e840d63459ca2__20210901_20210903', no_mirror),  # noqa: E501, contains file with bad size and sha256
     source('bigquery', 'datarepo-dev-96d8e08c', 'hca_dev_faeedcb0e0464be7b1ad80a3eeabb066__20210831_20210903'),
     # @formatter:on
 ]))

